rust/lance/src/index/append.rs

```diff
@@ -15,56 +15,73 @@
 use std::sync::Arc;
 
 use lance_core::{format::Index as IndexMetadata, Error, Result};
-use lance_index::scalar::lance_format::LanceIndexStore;
-use lance_index::IndexType;
-use log::info;
+use lance_index::{optimize::OptimizeOptions, scalar::lance_format::LanceIndexStore, IndexType};
 use roaring::RoaringBitmap;
 use snafu::{location, Location};
 use uuid::Uuid;
 
+use super::vector::ivf::optimize_vector_indices;
+use super::DatasetIndexInternalExt;
 use crate::dataset::scanner::ColumnOrdering;
 use crate::dataset::Dataset;
-use crate::index::vector::ivf::IVFIndex;
-
-use super::DatasetIndexInternalExt;
 
 /// Append new data to the index, without re-train.
 ///
 /// Returns the UUID of the new index along with a vector of newly indexed fragment ids
-pub async fn append_index(
+pub async fn append_index<'a>(
     dataset: Arc<Dataset>,
-    old_index: &IndexMetadata,
-) -> Result<Option<(Uuid, Option<RoaringBitmap>)>> {
-    let unindexed = dataset.unindexed_fragments(&old_index.name).await?;
-    if unindexed.is_empty() {
-        return Ok(None);
+    old_indices: &[&'a IndexMetadata],
+    options: &OptimizeOptions,
+) -> Result<Option<(Uuid, Vec<&'a IndexMetadata>, RoaringBitmap)>> {
+    if old_indices.is_empty() {
+        return Err(Error::Index {
+            message: "Append index: no prevoius index found".to_string(),
+            location: location!(),
+        });
     };
 
-    let frag_bitmap = old_index.fragment_bitmap.as_ref().map(|bitmap| {
-        let mut bitmap = bitmap.clone();
-        bitmap.extend(unindexed.iter().map(|frag| frag.id as u32));
-        bitmap
-    });
-
     let column = dataset
         .schema()
-        .field_by_id(old_index.fields[0])
+        .field_by_id(old_indices[0].fields[0])
         .ok_or(Error::Index {
             message: format!(
                 "Append index: column {} does not exist",
-                old_index.fields[0]
+                old_indices[0].fields[0]
             ),
             location: location!(),
         })?;
 
-    let index = dataset
-        .open_generic_index(&column.name, &old_index.uuid.to_string())
-        .await?;
+    let mut indices = Vec::with_capacity(old_indices.len());
+    for idx in old_indices {
+        let index = dataset
+            .open_generic_index(&column.name, &idx.uuid.to_string())
+            .await?;
+        indices.push(index);
+    }
```

Comment:
Again, could maybe use `Vec::from_iter` here?

Reply:
When I change to 

```
    let indices = stream::iter(old_indices.iter())
        .zip(repeat((dataset.clone(), col_name.clone())))
        .map(|(meta, (ds, col_name))| async move {
            ds.open_generic_index(&col_name, &meta.uuid.to_string())
                .await
        })
        .buffered(10)
        .try_collect::<Vec<_>>()
        .await?;
```

it complaints

```
error: higher-ranked lifetime error
   --> lance/src/index.rs:300:5
    |
300 |     #[instrument(skip_all)]
    |     ^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: could not prove `Pin<Box<{async block@lance/src/index.rs:300:5: 300:28}>>: CoerceUnsized<Pin<Box<(dyn futures::Future<Output = Result<(), lance_core::Error>> + std::marker::Send + 'i)>>>`
    = note: this error originates in the attribute macro `instrument` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `lance` (lib) due to previous error
```

Reply:
Ah, the return of the dreaded https://github.com/rust-lang/rust/issues/102211

Don't worry too much about it but I I've had luck with something like this...

```
    let indices = stream::iter(old_indices.iter())
        .zip(repeat((dataset.clone(), col_name.clone())))
        .map(|(meta, (ds, col_name))| async move {
            ds.open_generic_index(&col_name, &meta.uuid.to_string())
                .await
        })
        .collect::<Vec<_>>();
   let indices = stream::iter(indices)
        .buffered(10)
        .try_collect::<Vec<_>>()
        .await?;
```

...or you can just remove the call to `buffered` (that's the method that usually introduces the bogus error).